woocommerce/woocommerce-ios
# Product Details UI Mark1: Adds ProductLoaderVC + ProductDetailsVC

This PR adds the ability to launch the product details modal from either the `FulfillViewController` or `ProductListViewController`:

![product_loading](https://user-images.githubusercontent.com/154014/56066391-39770380-5d3d-11e9-9b94-d248e631c031.gif)

**Note:** everything in this PR has been feature flagged as `.localDeveloper`!

Fixes: #855 

### ProductLoaderViewController

The new `ProductLoaderViewController` (which is similar to our much-beloved `OrderLoaderViewController`) decouples the Product Detail VC from any specific navigation flow. ProductLoader does the following when launched:

**1.** Displays a fullscreen spinner and temp `Loading Product` nav bar title:
![product_load copy 2](https://user-images.githubusercontent.com/154014/56066942-ec942c80-5d3e-11e9-9392-f0ee1312ac85.png)

**2.** Asynchronously fetches a `Product` given the passed-in `productID` + `siteID`

**3.** When the product is retrieved, the (new) `ProductDetailsViewController` is instantiated and rendered in place

**4.** If an error occurs while loading the product, an error overlay is displayed:
![product_error copy 2](https://user-images.githubusercontent.com/154014/56066979-18171700-5d3f-11e9-81f8-3160d625bdf6.png)

### ProductDetailsViewController

The shell of `ProductDetailsViewController` was also implemented in this PR and it will be updated in subsequent PRs. For now this VC contains:
* Core Initialization logic (using real product data)
* Basic `UITableView` setup & config
* `EntityListener` setup & config
* Pull to refresh logic (including fetching real product data)

### Other stuff

* `ProductListViewController.swift` was located in the wrong project folder so it was moved
* Added a new `FeatureFlag` → `productDetails`

## Testing!
1. Take a look at the code focusing primarily on `ProductLoaderViewController` — `ProductDetailsViewController` will be updated in upcoming PRs
2. Run the unit tests, verify they are ✅ 
3. Build and run the app:
    * Verify you can see the shell product details screen when tapping on a order item in `FulfillViewController` or `ProductListViewController` 
    * Turn off your network connection and tap on an order line item in `FulfillViewController` or `ProductListViewController` — Verify the error overlay displays (eventually)
    * Test out pull-to-refresh on `ProductDetailsViewController`

@ctarda and @mindgraffiti → would you mind giving this PR a 👀 ?
